spanner/spanner_snippets/snippet_test.go

```diff
@@ -246,14 +288,66 @@ func TestSample(t *testing.T) {
 	assertContains(t, out, "4 Venue 4")
 	assertContains(t, out, "19 Venue 19")
 	assertContains(t, out, "42 Venue 42")
-
 	out = runCommand(t, "querywithqueryoptions", dbName)
 	assertContains(t, out, "4 Venue 4")
 	assertContains(t, out, "19 Venue 19")
 	assertContains(t, out, "42 Venue 42")
-
 	out = runCommand(t, "createclientwithqueryoptions", dbName)
 	assertContains(t, out, "4 Venue 4")
 	assertContains(t, out, "19 Venue 19")
 	assertContains(t, out, "42 Venue 42")
+
+	out = runBackupCommand(t, "createbackup", dbName, backupID)
+	assertContains(t, out, fmt.Sprintf("backups/%s", backupID))
+
+	out = runBackupCommand(t, "cancelbackup", dbName, cancelledBackupID)
+	assertContains(t, out, "Backup cancelled.")
+
+	out = runBackupCommand(t, "listbackups", dbName, backupID)
+	assertContains(t, out, fmt.Sprintf("/backups/%s", backupID))
+	assertContains(t, out, "Backups listed.")
+
+	out = runCommand(t, "listbackupoperations", dbName)
+	assertContains(t, out, fmt.Sprintf("on database %s", dbName))
+
+	out = runBackupCommand(t, "updatebackup", dbName, backupID)
+	assertContains(t, out, fmt.Sprintf("Updated backup %s", backupID))
+
+	out = runBackupCommand(t, "restorebackup", restoreDBName, backupID)
+	assertContains(t, out, fmt.Sprintf("Source database %s restored from backup", dbName))
+
+	WaitForDBReadyOptimizing(ctx, adminClient, restoreDBName, t)
+
+	// This command runs after a restore operation.
+	out = runCommand(t, "listdatabaseoperations", restoreDBName)
+	assertContains(t, out, fmt.Sprintf("Database %s restored from backup", restoreDBName))
+
+	out = runBackupCommand(t, "deletebackup", dbName, backupID)
+	assertContains(t, out, fmt.Sprintf("Deleted backup %s", backupID))
+}
+
+func WaitForDBReadyOptimizing(ctx context.Context, adminClient *database.DatabaseAdminClient, restoreDBName string, t *testing.T) {
+	// Wait for database to finish optimizing - cannot delete a backup if a database restored from it
```

Comment:
One of the Spanner team members said "Clients **should be able to delete databases in state READY_OPTIMIZING**, they **should be able to delete instances which have databases that are still optimizing**. Once restore is completed, you can delete databases and corresponding instances."

So we might be able to get rid of this to speed up the tests.

Reply:
I think this waiting is necessary. After I removed it, I got the following errors: 

```bash
TestBackupSample: snippet_test.go:356: run("deletebackup", "projects/e2e-debugging/instances/hengfeng-test-instance/databases/test-golang-samples-tests-7")
: rpc error: code = FailedPrecondition desc = Backup projects/e2e-debugging/instances/hengfeng-test-instance/backups/backup-golang-samples-tests-7 cannot be del
eted at the moment because it is still in use by 1 databases for pending restore activity. Please try deleting the backup once the restore or post-restore optim
ize operations have completed on these databases. The databases using this backup are displayed in the UI in the Backups table `In use by` column, and returned
by the GetBackup API.

    TestBackupSample: snippet_test.go:357: got output "deletebackup failed with rpc error: code = FailedPrecondition desc = Backup projects/e2e-debugging/instan
ces/hengfeng-test-instance/backups/backup-golang-samples-tests-7 cannot be deleted at the moment because it is still in use by 1 databases for pending restore a
ctivity. Please try deleting the backup once the restore or post-restore optimize operations have completed on these databases. The databases using this backup
are displayed in the UI in the Backups table `In use by` column, and returned by the GetBackup API."; want it to contain "Deleted backup backup-golang-samples-t
ests-7"
```

Reply:
Oh no, ok I'll add you to the thread offline about this.